// leetcode-query/lib/index.js
"use strict";var j=Object.create;var S=Object.defineProperty;var V=Object.getOwnPropertyDescriptor;var H=Object.getOwnPropertyNames;var J=Object.getPrototypeOf,W=Object.prototype.hasOwnProperty;var Y=(n,t)=>{for(var e in t)S(n,e,{get:t[e],enumerable:!0})},L=(n,t,e,i)=>{if(t&&typeof t=="object"||typeof t=="function")for(let s of H(t))!W.call(n,s)&&s!==e&&S(n,s,{get:()=>t[s],enumerable:!(i=V(t,s))||i.enumerable});return n};var g=(n,t,e)=>(e=n!=null?j(J(n)):{},L(t||!n||!n.__esModule?S(e,"default",{value:n,enumerable:!0}):e,n)),K=n=>L(S({},"__esModule",{value:!0}),n);var ie={};Y(ie,{BASE_URL:()=>d,BASE_URL_CN:()=>f,Cache:()=>v,Credential:()=>y,LeetCode:()=>_,LeetCodeCN:()=>G,Mutex:()=>k,RateLimiter:()=>b,USER_AGENT:()=>c,cache:()=>h,caches:()=>X,default:()=>te});module.exports=K(ie);var M=g(require("events")),P=g(require("node-fetch"));var v=class{constructor(){this._table={}}get(t){let e=this._table[t];if(e){if(e.expires>Date.now())return e.value;this.remove(t)}return null}set(t,e,i=6e4){this._table[t]={key:t,value:e,expires:i>0?Date.now()+i:0}}remove(t){delete this._table[t]}clear(){this._table={}}load(t){this._table=JSON.parse(t)}},h=new v,X={default:h};var d="https://leetcode.com",f="https://leetcode.cn",c="Mozilla/5.0 LeetCode API";var U=g(require("node-fetch"));function p(n){return n.split(";").map(t=>t.trim().split("=")).reduce((t,e)=>(t[e[0]]=e[1],t),{})}async function Z(){let n=await(0,U.default)(d,{headers:{"user-agent":c}}).then(e=>e.headers.get("set-cookie"));return p(n).csrftoken}var y=class{constructor(t){t&&(this.session=t.session,this.csrf=t.csrf)}async init(t){return this.csrf=await Z(),t&&(this.session=t),this}};var I=`
query ($username: String!) {
    userContestRanking(username: $username) {
        attendedContestsCount
        rating
        globalRanking
        totalParticipants
        topPercentage
        badge {
            name
        }
    }
    userContestRankingHistory(username: $username) {
        attended
        trendDirection
        problemsSolved
        totalProblems
        finishTimeInSeconds
        rating
        ranking
        contest {
            title
            startTime
        }
    }
}
`;var T=`
query {
    activeDailyCodingChallengeQuestion {
        date
        link
        question {
            questionId
            questionFrontendId
            boundTopicId
            title
            titleSlug
            content
            translatedTitle
            translatedContent
            isPaidOnly
            difficulty
            likes
            dislikes
            isLiked
            similarQuestions
            exampleTestcases
            contributors {
                username
                profileUrl
                avatarUrl
            }
            topicTags {
                name
                slug
                translatedName
            }
            companyTagStats
            codeSnippets {
                lang
                langSlug
                code
            }
            stats
            hints
            solution {
                id
                canSeeDetail
                paidOnly
                hasVideoSolution
                paidOnlyVideo
            }
            status
            sampleTestCase
            metaData
            judgerAvailable
            judgeType
            mysqlSchemas
            enableRunCode
            enableTestMode
            enableDebugger
            envInfo
            libraryUrl
            adminUrl
            challengeQuestion {
                id
                date
                incompleteChallengeCount
                streakCount
                type
            }
            note
        }
    }
}
`;var N=`
query ($titleSlug: String!) {
    question(titleSlug: $titleSlug) {
        questionId
        questionFrontendId
        boundTopicId
        title
        titleSlug
        content
        translatedTitle
        translatedContent
        isPaidOnly
        difficulty
        likes
        dislikes
        isLiked
        similarQuestions
        exampleTestcases
        contributors {
            username
            profileUrl
            avatarUrl
        }
        topicTags {
            name
            slug
            translatedName
        }
        companyTagStats
        codeSnippets {
            lang
            langSlug
            code
        }
        stats
        hints
        solution {
            id
            canSeeDetail
            paidOnly
            hasVideoSolution
            paidOnlyVideo
        }
        status
        sampleTestCase
        metaData
        judgerAvailable
        judgeType
        mysqlSchemas
        enableRunCode
        enableTestMode
        enableDebugger
        envInfo
        libraryUrl
        adminUrl
        challengeQuestion {
            id
            date
            incompleteChallengeCount
            streakCount
            type
        }
        note
    }
}
`;var E=`
query ($categorySlug: String, $limit: Int, $skip: Int, $filters: QuestionListFilterInput) {
    problemsetQuestionList: questionList( 
        categorySlug: $categorySlug
        limit: $limit
        skip: $skip
        filters: $filters
    ) {
        total: totalNum
        questions: data {
            acRate 
            difficulty 
            freqBar 
            questionFrontendId 
            isFavor
            isPaidOnly 
            status 
            title 
            titleSlug 
            topicTags { name id slug } 
            hasSolution 
            hasVideoSolution
        }
    }
}
`;var D=`
query ($username: String!) {
    allQuestionsCount {
        difficulty
        count
    }
    matchedUser(username: $username) {
        username
        socialAccounts
        githubUrl
        contributions {
            points
            questionCount
            testcaseCount
        }
        profile {
            realName
            websites
            countryName
            skillTags
            company
            school
            starRating
            aboutMe
            userAvatar
            reputation
            ranking
        }
        submissionCalendar
        submitStats {
            acSubmissionNum {
                difficulty
                count
                submissions
            }
            totalSubmissionNum {
                difficulty
                count
                submissions
            }
        }
        badges {
            id
            displayName
            icon
            creationDate
        }
        upcomingBadges {
            name
            icon
        }
        activeBadge {
            id
        }
    }
    recentSubmissionList(username: $username, limit: 20) {
        title
        titleSlug
        timestamp
        statusDisplay
        lang
    }
}
`;var Q=`
query ($username: String!, $limit: Int) {
    recentSubmissionList(username: $username, limit: $limit) {
        title
        titleSlug
        timestamp
        statusDisplay
        lang
    }
}
`;var A=`
query ($offset: Int!, $limit: Int!, $slug: String) {
    submissionList(offset: $offset, limit: $limit, questionSlug: $slug) {
        hasNext
        submissions { id lang time timestamp statusDisplay runtime url isPending title memory titleSlug }
    }
}
`;var $=`
query {
    userStatus {
      userId
      username
      avatar
      isSignedIn
      isMockUser
      isPremium
      isAdmin
      isSuperuser
      isTranslator
      permissions
    }
  }
  `;var O=g(require("events")),k=class extends O.default{constructor(e=1){super();this.space=e,this.used=0,this.releases=[]}async lock(){if(this.used>=this.space){let e=new Promise(i=>this.releases.push(i));this.emit("wait",{lock:e,release:this.releases[this.releases.length-1]}),await e}return this.used++,this.emit("lock"),this.used}unlock(){var e;return this.used<=0?0:(this.releases.length>0&&((e=this.releases.shift())==null||e()),this.used--,this.emit("unlock"),this.used<=0&&this.emit("all-clear"),this.used)}resize(e){var i;for(this.space=e;this.used<e&&this.releases.length>0;)(i=this.releases.shift())==null||i();return this.space}full(){return this.used>=this.space}waiting(){return this.releases.length}},b=class extends k{constructor({limit:e=20,interval:i=1e4,concurrent:s=2}={}){super(s);this.count=0;this.last=0;this.time_mutex=new k(e),this.interval=i,this.time_mutex.on("lock",(...r)=>this.emit("time-lock",...r)),this.time_mutex.on("unlock",(...r)=>this.emit("time-unlock",...r))}async lock(){return this.last+this.interval<Date.now()?this.reset():this.time_mutex.full()&&!this.timer&&this.cleaner(),await this.time_mutex.lock(),this.count++,super.lock()}reset(){for(;this.count>0;)this.time_mutex.unlock(),this.count--;this.last=Date.now(),this.emit("timer-reset")}cleaner(){this.timer=setTimeout(()=>{this.reset(),setTimeout(()=>{this.time_mutex.waiting()>0?this.cleaner():this.timer=void 0},0)},this.last+this.interval-Date.now())}set limit(e){this.time_mutex.resize(e)}};var R=class extends M.default{constructor(e=null,i=h){super();this.limiter=new b;let s;this.initialized=new Promise(r=>{s=r}),this.cache=i,e?(this.credential=e,setImmediate(()=>s())):(this.credential=new y,this.credential.init().then(()=>s()))}async user(e){await this.initialized;let{data:i}=await this.graphql({variables:{username:e},query:D});return i}async user_contest_info(e){await this.initialized;let{data:i}=await this.graphql({variables:{username:e},query:I});return i}async recent_submissions(e,i=20){await this.initialized;let{data:s}=await this.graphql({variables:{username:e,limit:i},query:Q});return s.recentSubmissionList||[]}async submissions({limit:e=20,offset:i=0,slug:s}={}){await this.initialized;let r=[],m=new Set,l=i;for(;r.length<e;){let{data:u}=await this.graphql({variables:{offset:l,limit:e-r.length>20?20:e-r.length,slug:s},query:A});for(let o of u.submissionList.submissions)o.id=parseInt(o.id,10),o.timestamp=parseInt(o.timestamp,10)*1e3,o.isPending=o.isPending!=="Not Pending",o.runtime=parseInt(o.runtime,10)||0,o.memory=parseFloat(o.memory)||0,!m.has(o.id)&&(m.add(o.id),r.push(o));if(!u.submissionList.hasNext)break;l+=20}return r}async submission(e){var i;await this.initialized;try{await this.limiter.lock();let m=(i=(await(await(0,P.default)(`${d}/submissions/detail/${e}/`,{headers:{origin:d,referer:d,cookie:`csrftoken=${this.credential.csrf||""}; LEETCODE_SESSION=${this.credential.session||""};`,"user-agent":c}})).text()).match(/var pageData = ({[^]+?});/))==null?void 0:i[1],l=new Function("return "+m)(),u={id:parseInt(l.submissionId),problem_id:parseInt(l.questionId),runtime:parseInt(l.runtime),runtime_distribution:l.runtimeDistributionFormatted?JSON.parse(l.runtimeDistributionFormatted).distribution.map(o=>[+o[0],o[1]]):[],runtime_percentile:0,memory:parseInt(l.memory),memory_distribution:l.memoryDistributionFormatted?JSON.parse(l.memoryDistributionFormatted).distribution.map(o=>[+o[0],o[1]]):[],memory_percentile:0,code:l.submissionCode,details:l.submissionData};return u.runtime_percentile=u.runtime_distribution.reduce((o,[x,w])=>o+(x>=u.runtime?w:0),0),u.memory_percentile=u.memory_distribution.reduce((o,[x,w])=>o+(x>=u.memory/1e3?w:0),0),this.limiter.unlock(),u}catch(s){throw this.limiter.unlock(),s}}async problems({category:e="",offset:i=0,limit:s=100,filters:r={}}={}){await this.initialized;let m={categorySlug:e,skip:i,limit:s,filters:r},{data:l}=await this.graphql({variables:m,query:E});return l.problemsetQuestionList}async problem(e){await this.initialized;let{data:i}=await this.graphql({variables:{titleSlug:e.toLowerCase().replace(/\s/g,"-")},query:N});return i.question}async daily(){await this.initialized;let{data:e}=await this.graphql({query:T});return e.activeDailyCodingChallengeQuestion}async whoami(){await this.initialized;let{data:e}=await this.graphql({operationName:"globalData",variables:{},query:$});return e.userStatus}async graphql(e){await this.initialized;try{await this.limiter.lock();let i=d,s=await(0,P.default)(`${i}/graphql`,{method:"POST",headers:{"content-type":"application/json",origin:i,referer:i,cookie:`csrftoken=${this.credential.csrf||""}; LEETCODE_SESSION=${this.credential.session||""};`,"x-csrftoken":this.credential.csrf||"","user-agent":c},body:JSON.stringify(e)});if(this.emit("receive-graphql",s.clone()),s.headers.has("set-cookie")){let r=p(s.headers.get("set-cookie"));r.csrftoken&&(this.credential.csrf=r.csrftoken,this.emit("update-csrf",this.credential))}return this.limiter.unlock(),s.json()}catch(i){throw this.limiter.unlock(),i}}},_=R;var B=g(require("events")),z=g(require("node-fetch"));var F=g(require("node-fetch"));async function ee(){let t=(await(0,F.default)(`${f}/graphql/`,{method:"POST",headers:{"content-type":"application/json","user-agent":c},body:JSON.stringify({operationName:"nojGlobalData",variables:{},query:`query nojGlobalData {
  siteRegion
  chinaHost
  websocketUrl
}
`})})).headers.get("set-cookie");return p(t).csrftoken}var C=class{constructor(t){t&&(this.session=t.session,this.csrf=t.csrf)}async init(t){return this.csrf=await ee(),t&&(this.session=t),this}};var q=class extends B.default{constructor(e=null,i=h){super();this.limiter=new b;let s;this.initialized=new Promise(r=>{s=r}),this.cache=i,e?(this.credential=e,setImmediate(()=>s())):(this.credential=new C,this.credential.init().then(()=>s()))}async user(e){await this.initialized;let{data:i}=await this.graphql({operationName:"getUserProfile",variables:{username:e},query:`
            query getUserProfile($username: String!) {
                userProfileUserQuestionProgress(userSlug: $username) {
                    numAcceptedQuestions { difficulty count }
                    numFailedQuestions { difficulty count }
                    numUntouchedQuestions { difficulty count }
                }
                userProfilePublicProfile(userSlug: $username) {
                    username haveFollowed siteRanking
                    profile { 
                        userSlug realName aboutMe userAvatar location gender websites skillTags contestCount asciiCode
                        medals { name year month category }
                        ranking { 
                            currentLocalRanking currentGlobalRanking currentRating totalLocalUsers totalGlobalUsers
                        }
                        socialAccounts { provider profileUrl }
                    }
                }
            }
            `});return i}async graphql(e,i="/graphql/"){await this.initialized;try{await this.limiter.lock();let s=f,r=await(0,z.default)(`${s}${i}`,{method:"POST",headers:{"content-type":"application/json",origin:s,referer:s,cookie:`csrftoken=${this.credential.csrf||""}; LEETCODE_SESSION=${this.credential.session||""};`,"x-csrftoken":this.credential.csrf||"","user-agent":c},body:JSON.stringify(e)});if(this.emit("receive-graphql",r.clone()),r.headers.has("set-cookie")){let m=p(r.headers.get("set-cookie"));m.csrftoken&&(this.credential.csrf=m.csrftoken,this.emit("update-csrf",this.credential))}return this.limiter.unlock(),r.json()}catch(s){throw this.limiter.unlock(),s}}},G=q;var te=_;0&&(module.exports={BASE_URL,BASE_URL_CN,Cache,Credential,LeetCode,LeetCodeCN,Mutex,RateLimiter,USER_AGENT,cache,caches});
